torchgeo/datasets/copernicus/base.py

```diff
@@ -0,0 +1,249 @@
+            transforms: A function/transform that takes input sample and its target as
+                entry and returns a transformed version.
+            download: If True, download dataset and store it in the root directory.
+            checksum: If True, check the MD5 of the downloaded files (may be slow).
+
+        Raises:
+            DatasetNotFoundError: If dataset is not found and *download* is False.
+        """
+        self.root = root
+        self.split = split
+        self.bands = bands or self.all_bands
+        self.band_indices = [self.all_bands.index(i) + 1 for i in self.bands]
+        self.transforms = transforms
+        self.download = download
+        self.checksum = checksum
+
+        self._verify()
+
+        filepath = os.path.join(root, self.directory, self.filename.format(split))
+        self.files = pd.read_csv(filepath, header=None)[0]
+
+    def __len__(self) -> int:
+        """Return the length of the dataset.
+
+        Returns:
+            Length of the dataset.
+        """
+        return len(self.files)
+
+    def __getitem__(self, index: int) -> dict[str, Tensor]:
+        """Return an index within the dataset.
+
+        Args:
+            index: Index to return.
+
+        Returns:
+            Data and labels at that index.
+        """
+        sample = self._load_image(index) | self._load_target(index)
+
+        if self.transforms is not None:
+            sample = self.transforms(sample)
+
+        return sample
+
+    @abstractmethod
+    def _load_image(self, index: int) -> dict[str, Tensor]:
+        """Load an image.
+
+        Args:
+            index: Index to return.
+
+        Returns:
+            An image sample.
+        """
+
+    @abstractmethod
+    def _load_target(self, index: int) -> dict[str, Tensor]:
+        """Load a target label or mask.
+
+        Args:
+            index: Index to return.
+
+        Returns:
+            A target sample.
+        """
+
+    def _verify(self) -> None:
+        """Verify the integrity of the dataset."""
+        # Check if the files already exist
+        path = os.path.join(self.root, self.directory, self.filename.format(self.split))
+        if os.path.exists(path):
+            return
+
+        # Check if the zip file already exists (if so then extract)
+        if os.path.exists(os.path.join(self.root, self.zipfile)):
+            self._extract()
+            return
+
+        # Check if the user requested to download the dataset
+        if not self.download:
+            raise DatasetNotFoundError(self)
+
+        # Download and extract the dataset
+        self._download()
+
+    def _extract(self) -> None:
+        """Extract the dataset."""
+        extract_archive(os.path.join(self.root, self.zipfile))
+
+    def _download(self) -> None:
+        """Download the dataset."""
+        md5 = self.md5 if self.checksum else None
+        download_and_extract_archive(self.url, self.root, md5=md5)
+
+    def plot(
```

Comment:
I would actually like to move this idea to NonGeoDataset and GeoDataset in the future. 3/4 of our custom plotting methods could be removed by assuming that `all_bands`, `rgb_bands`, and `cmap` are defined as class attributes, using percentile normalization, and assuming standard sample key names. Somewhat relies on #985 though.